project_timesheet_holidays_contract/models/hr_employee.py

```diff
@@ -0,0 +1,184 @@
+# Copyright 2024 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+from datetime import datetime
+
+from odoo import _, fields, models
+from odoo.exceptions import ValidationError
+
+
+class HrEmployee(models.Model):
+    _inherit = "hr.employee"
+
+    def _get_calendar(self, date_from=None):
+        if not date_from:  # no `date_from` => call `super()`
+            return super()._get_calendar(date_from)
+
+        date_from = (
+            date_from if not isinstance(date_from, datetime) else date_from.date()
+        )
+        date_to = self.env.context.get("date_to", None)
+        if date_to is None:
+            date_to = fields.Date.today()
+        date_to = date_to if not isinstance(date_to, datetime) else date_to.date()
+
+        contracts = (
+            self.sudo()
+            .with_context(active_test=True)
+            .env["hr.contract"]
+            .search(
+                [
+                    ("employee_id", "=", self.id),
+                    ("date_start", "<=", date_from),
+                    "|",
+                    ("date_end", ">=", date_to),
+                    ("date_end", "=", False),
+                ]
+            )
+        )
+
+        # If only one contract, return it regardless of its state
+        if len(contracts) == 1:
+            return contracts[0].resource_calendar_id
+
+        # Filter out cancelled contracts if there are multiple contracts
+        valid_contracts = contracts.filtered(lambda c: c.state != "cancel")
```

Comment:
So in case there's only one contract that is canceled, it still must be used? :raised_eyebrow: